### PR TITLE
Move threadCount to Configuration

### PR DESCRIPTION
**Pull request by theofidry**
As I was looking into #1150, I was blocked by this question: where should this parameter go? Should it stored in the `Configuration` be part of service constructor(s) or passed around as a parameter?

If we take a closer look at Infection, the reality is as follows:

- we have a lot of parameters (see `Configuration`)
- most services depends on those parameters (i.e. depends on `Configuration`)

And it totally makes sense. As much as `Container::withDynamicParameters()` might be a bit confusing at first, it's neither an out of ordinary pattern and neither a bad fit: the alternative means passing the config to 3/4 of our services and that does not make any sense.

So back to the initial question. I think the best way to approach this is to ask ourselves: is that parameter an implementation detail of this class, or a generic parameter? For better clarity let's take our example here from this PR:

```php
// Prior to this PR
class ParallelProcessRunner {
    function run($processes, $threadCount)
}
```

`ParallelProcessRunner` is a concrete class without any parents. But if we imagined it had an interface, e.g. `ProcessRunner`, do the parameters `$processes` and `$threadCount` have a place there or are they an implementation detail of `ParallelProcessRunner`? IMO `$process` has a place yes, but not `$threadCount`.

Another example: `MutationTestingResultsLogger`:

```php
interface MutationTestingResultsLogger
{
    /**
     * Logs results of Mutation Testing to somewhere
     */
    public function log(): void;
}
```

Would it make sense to have a config object passed to `log()` for each type of logger? IMO while doable, it does not really makes sense either here (it could in other contexts maybe?).

I might revisit some similar cases in the future, but for now I'm fixing this one and I will add the `dryRun` parameter to `Configuration` in a similar fashion

### CODE REVIEW

File: tests/phpunit/Process/Runner/Parallel/ParallelProcessRunnerTest.php [hunk 1]
```diff
@@ -36,148 +36,175 @@
...
-        $runner = new ParallelProcessRunner(self::makeEventAccounter($eventDispatcher));
-        $runner->run($processes, 4, 0);
+        $runner->run($processes);
     }
 
-    public function test_it_checks_for_timeout(): void
+    public function test_it_checks_if_the_executed_processes_time_out(): void
     {
-        $processes = (function (): iterable {
+        $processes = (function (): Generator {
```
Comment (sanmai) [inline tests/phpunit/Process/Runner/Parallel/ParallelProcessRunnerTest.php]: iterable?

File: tests/phpunit/Fixtures/Process/DummyProcessBearer.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    public function markAsTimedOut(): void
+    {
+        if (!$this->expectTimeOut) {
+            Assert::fail(sprintf(
+                'Did not expect "%s()" to be called',
+                __FUNCTION__
+            ));
+        }
+    }
+}
```
Comment (sanmai) [inline tests/phpunit/Fixtures/Process/DummyProcessBearer.php]: Easier than a mock?

Comment (theofidry) [reply tests/phpunit/Fixtures/Process/DummyProcessBearer.php]: I think so. In general I don't lick mocks that much, if they are appropriate why not, but if you can easily do without I prefer without. It has several advantages:

- Rots less easily
- Easier to refactor
- Less magic (so more accessible)
- Faster (mocking is slow)

I also got the impression that [jakzal](https://github.com/jakzal), [frankdejonge](https://github.com/frankdejonge) & [sebastian](https://github.com/sebastianbergmann) recommend the same. I would not say as far as blindly do without mocks at all, but I'm considering hunting down some cases where I feel we should not need mock

Comment (sanmai) [reply tests/phpunit/Fixtures/Process/DummyProcessBearer.php]: I'm not a big fan of mocks tbh, so it's just surprising to see a different approach for once.

**Pull request merged**